dev/spark_overview.md

```diff
@@ -0,0 +1,452 @@
+finish.
+
+This section explains some common problems.
+
+### spark.sql.shuffle.partitions
+The most common performance issue we encounter when running complex queries is due to a non-ideal
+setting for `spark.sql.shuffle.partitions`. The default Spark value is 200. Some online sources
+recommend setting it to 1-4x the total number of CPUs in your cluster. For example, if you have two
+Eagle nodes, 72 - 288. This [video](https://www.youtube.com/watch?v=daXEp4HmS-E&t=4251s) by a Spark
+developer offers a recommendation that has worked out better.
+
+Use this formula:
+```
+num_partitions = max_shufffle_write_size / target_partition_size
+```
+You will have to run your job once to determine `max_shuffle_write_size`. You can find it on the
+Spark UI Stages tab in the Shuffle Write column. Your `target_partition_size` should be between
+128 - 200 MB.
+
+The minimum partitions value should be the total number of cores in the cluster unless you want to
+leave some cores available for other jobs that may be running simultaneously.
+
+### Dynamic allocation
+This feature is disabled by default on standalone clusters. It is described
+[here](https://spark.apache.org/docs/latest/job-scheduling.html#dynamic-resource-allocation).
+This section does not necessarily line up with the Spark documentation and will change as we learn
+more.
+
+We have observed two cases where enabling dynamic allocation on a standalone cluster significantly
+improves performance:
+
+- When a Spark job produces a huge number of tasks. This can happen when checking dimension
+associations as well as queries that map a dataset to project dimensions.
+- When there is really only enough work for one executor but Spark distributes it to all executors
+anyway. The inter-process communication adds tons of overhead. The executors also appear to do a
+lot more work. With dynamic allocation Spark gradually adds executors and these problems don't
+always occur.
+
+The second issue sometimes occurs when submitting a dataset to a project in the registry.
+We recommend enabling dynamic allocation when doing this. If that is enabled then dsgrid code will
+reconfigure the SparkSession to use a single executor with one core.
+
+Set these values in your `spark-defaults.conf`. `spark.shuffle.service.enabled` must be set before
+you start the workers.
+```
+spark.dynamicAllocation.enabled true
+spark.dynamicAllocation.shuffleTracking.enabled true
+spark.shuffle.service.enabled true
+```
+
+### Slow local storage
+Spark will write lots of temporary data to local storage during shuffle operations. If your joins
+cause lots of shuffling, it is very important that your local storage be fast. If you use an Eagle
+node with a slow spinning disk (130 MB/s), your job will likelly fail. Use a `bigmem` or `gpu`
+node. They have SSDs that can write data at 2 GB/s.
+
+We have not observed any downside to having this feature enabled.
+
+### Too many executors are involved in a job
```

Comment:
This seems to go more hand in hand with dynamic allocation, suggests moving it close to that section.